src/net.cpp

```diff
@@ -558,6 +560,15 @@ Network CNode::ConnectedThroughNetwork() const
     return IsInboundConn() && m_inbound_onion ? NET_ONION : addr.GetNetClass();
 }
 
+void CNode::UpdateConnectionType(ConnectionType new_conn_type)
+{
+    assert(m_conn_type == ConnectionType::INBOUND);
+    assert(new_conn_type == ConnectionType::INBOUND_BLOCK_RELAY);
+    m_conn_type = new_conn_type;
+    // We can now drop some unnecessary data structures.
+    // TODO: deallocate m_tx_relay and m_addr_known.
```

Comment:
I am confused what constitutes a new connection type v/s a negotiated feature of a connection. Other negotiated features that do not seem to change the connection type include `WTXIDRELAY`,`SENDADDRV2`, `SENDCMPCT`, etc.

As a starting point, I'd suggest that connection types imply a different level of trust in the peer (eg. `INBOUND` vs `OUTBOUND`) or connection limits(eg. 2 for `OUTBOUND_BLOCK_RELAY`, 1 for `FEELER`) and all other features are attributes of a connection.  This also opens up future possibilities for peers that want other combinations like block and addr relay, but no tx relay, etc.

Reply:
I wondered whether the better approach would be to add a new bool to indicate whether tx-relay is disabled (like we have for compact blocks, or sendheaders, etc), or to update the connection type like I've done here. 

My hesitation to add a new bool was that it seemed counter to the recent work to refactor `net` to use connection types as an alternative to the bucket-of-bool's approach for deciding how to interact with a peer.  If we have a set of semantics that we understand go together for a given peer, it's nice for code readability to encode that as simply as possible.

On the other hand, my intention with the BIP was to allow future flexibility for what semantics might be possible (eg my intention is to allow for addr-relay to happen in the future if we add an explicit way to negotiate addr relay, as an example), so from that perspective, having a connection type perhaps implies too much of an understanding of what the peer may be doing.  And for inbound peers in particular, we never really know what they're doing, so perhaps having more than just "inbound" as a connection type may be misleading to future code readers as well.

I like the idea of moving initialization of `m_tx_relay` to VERACK, but not sure if I'll end up tackling that here or deferring for the current refactoring work in progress to make more headway (I think we need to add a lock that guards access to the `m_tx_relay` pointer itself?)

Reply:
I think we've got a few different ways in which nodes differ:

 1. which random features they support (wtxid relay, compact blocks, addrv2; erlay, pkg relay, p2p encryption in the future) that might make things more efficient (and which we might prefer our connections to support) but don't really change anything
 2. how much we trust them (manually privileged/selected, we chose them as an outbound, they chose us as an inbound)
 3. why we've got the connection (manual setup for some higher level purpose, tx relay, block relay, feeler, addrfetch)
 4. which limits we apply to them (8 outbounds, 2 outbound block relay, 115 (?) inbounds)

We should use bools instead of conn_type for random features (otherwise we either get an exponential explosion of types or have to sequence features), and I don't think trust/"why" really affects inbounds that only wants blocks or not, but it could make sense to have a new limit imply a new connection-type? Given the limits overlap a little (extra outbounds vs feelers), and what to do when you have extra peers requires looking at other data anyway, I'm not sure that's very convincing though.

Hmm, how much we "trust" a connection might change over the course of a connection in future if we allow random inbound peers to authenticate themselves to us... So I guess that leaves me thinking that we should have `const m_conn_type` representing our purpose in having the connection? (Different purposes meaning substantial changes in the protocol state machine -- eg, blocks only we deliberately opt out of tx and addr relay even if we otherwise support both, feelers we disconnect after verack, addrfetch we disconnect after getting addresses, inbounds we have different delays for and send VERSION later)

> (I think we need to add a lock that guards access to the m_tx_relay pointer itself?)

Maybe? I think there are already way more per-peer mutexes than are really useful...